js/src/common/states/PaginatedListState.ts

```diff
@@ -0,0 +1,211 @@
+
+  /**
+   * Load a new page of results.
+   */
+  protected loadPage(page = 1): Promise<T[]> {
+    const params = this.requestParams();
+    params.page = { offset: this.pageSize * (page - 1) };
+
+    if (Array.isArray(params.include)) {
+      params.include = params.include.join(',');
+    }
+
+    return app.store.find(this.type, params);
+  }
+
+  /**
+   * Get the parameters that should be passed in the API request.
+   * Do not include page offset unless subclass overrides loadPage.
+   *
+   * @abstract
+   * @see loadPage
+   */
+  protected requestParams(): any {
+    return this.params;
+  }
+
+  /**
+   * Update the `this.params` object, calling `refresh` if they have changed.
+   * Use `requestParams` for converting `this.params` into API parameters
+   *
+   * @param newParams
+   * @param page
+   * @see requestParams
+   */
+  public refreshParams(newParams, page: number) {
+    if (this.isEmpty() || this.paramsChanged(newParams)) {
+      this.params = newParams;
+
+      return this.refresh(page);
+    }
+  }
+
+  public refresh(page: number = 1) {
+    this.initialLoading = true;
+    this.loadingPrev = false;
+    this.loadingNext = false;
+
+    m.redraw();
+
+    this.location = { page };
+
+    return this.loadPage()
+      .then((results: T[]) => {
+        this.pages = [];
+        this.parseResults(this.location.page, results);
+      })
+      .finally(() => (this.initialLoading = false));
+  }
+
+  public getPages() {
+    return this.pages;
+  }
+  public getLocation(): PaginationLocation {
+    return this.location;
+  }
+
+  public isLoading(): boolean {
+    return this.initialLoading || this.loadingNext;
+  }
+  public isInitialLoading(): boolean {
+    return this.initialLoading;
+  }
+  public isLoadingPrev(): boolean {
+    return this.loadingPrev;
+  }
+  public isLoadingNext(): boolean {
+    return this.loadingNext;
+  }
+
+  public hasItems(): boolean {
+    return !!this.getAllItems().length;
+  }
+  public isEmpty(): boolean {
+    return !this.isInitialLoading() && !this.hasItems();
+  }
+
+  public hasPrev(): boolean {
+    return this._hasPrev;
+  }
+  public hasNext(): boolean {
+    return this._hasNext;
+  }
+
+  /**
+   * Stored state parameters.
+   */
+  public getParams(): any {
+    return this.params;
+  }
+
+  protected getNextPageNumber(): number {
+    const pg = this.pages[this.pages.length - 1]?.number;
+
+    if (pg && !isNaN(pg)) {
+      return pg + 1;
+    } else {
+      return this.location.page;
+    }
+  }
+  protected getPrevPageNumber(): number {
+    const pg = this.pages[0]?.number;
+
+    if (pg && !isNaN(pg)) {
+      // If the calculated page number is less than 1,
+      // return 1 as the prev page (first possible page number)
+      return Math.max(pg - 1, 1);
+    } else {
+      return this.location.page;
+    }
+  }
+
+  protected paramsChanged(newParams): boolean {
+    return Object.keys(newParams).some((key) => this.requestParams()[key] !== newParams[key]);
```

Comment:
We should also check params.filter and params.qBuilder

Reply:
Scap qBuilder as per the comment above, but `filter` should be checked.

Reply:
Wouldn't we want to use e.g. lodash's `isEqual` function for this, and not just check specific arguments? 

Disregard the fact that we removed lodash from core recently 😂

Reply:
Not sure. The way it is right now, IIRC filter is assembled in `requestParams`, other than that it's a shallow object. Question is whether we want to keep it as shallow (in which case this comment is unnecessary).

Reply:
Hm this code also assumes `this.params` is in the format of `this.requestParams()`... wouldn't that most likely not be the case? As we have the params method + `requestParams` for the actual request? 🤔

Reply:
`requestParams` MUST be JSON API format, but `params` can be a weird intermediate format. We could go either way on this, but I think an intermediate format might make sense, as extensions can add data to it, and then modify how that data is serialized through `requestParams`.

Reply:
Don't we have an intermediate format now? With `params` and `getParams` - tho perhaps we should rename `getParams` as that doesn't sound like a method you'd want to extend in a subclass. 

And we can't really set `this.params` to the output of the intermediate format cause it could end up in things added over and over...

Unless I'm overthinking this.